Cherry-pick vir! macro from #54

dario23 opened the pull request.

dario23 committed:
Cherry-pick vir macro from PR #54

see https://github.com/viperproject/prusti-dev/pull/54 and
https://github.com/viperproject/prusti-dev/pull/54/commits/4c59d52f0e1c94ca4014a48f74cf13181717aab2

dario23 committed:
Some unit tests for `vir!` macro

Aurel300 commented on prusti-common/src/vir/vir_macro.rs:
```diff
@@ -0,0 +1,125 @@
+        crate::vir::Expr::BinOp(vir::BinOpKind::Or,
+            Box::new(vir!($head)),
+            Box::new(vir!($tail)),
+            crate::vir::Position::default())
```
I think there are functions defined in `Expr` that build particular binary operations with a default position, would make this a bit shorter. There might be similar for `Stmt`s too?

dario23 replied:
some `Stmt`s have, some don't .. i don't see why, exactly. tidied up a bit.

Aurel300 commented on prusti-common/src/vir/vir_macro.rs:
```diff
@@ -0,0 +1,125 @@
+    };
+    ($antecedent: tt ==> $consequent: tt) => {
+        crate::vir::Expr::BinOp(vir::BinOpKind::Implies,
+            Box::new(vir!($antecedent)),
```
If there is no "shorthand constructor" as described above, use `box xyz` syntax for the boxes.

Aurel300 commented on prusti-common/src/vir/vir_macro.rs:
```diff
@@ -0,0 +1,125 @@
+    (forall $($name: ident : $type: ident),+ :: {$trigger: tt} $body: tt) => {
+        crate::vir::Expr::ForAll(
+            vec![$(vir_local!($name: $type)),+],
+            vec![crate::vir::Trigger::new(vec![vir!($trigger)])],
```
There can be multiple trigger sets, each with multiple expressions in it. I think covering the former case is more important.
```suggestion
    (forall $($name: ident : $type: ident),+ :: { $($triggers: tt),* } $body: tt) => {
        crate::vir::Expr::ForAll(
            vec![$(vir_local!($name: $type)),+],
            vec![$(crate::vir::Trigger::new(vec![vir!($triggers)])),+],
```

Aurel300 commented on prusti-common/src/vir/vir_macro.rs:
```diff
@@ -0,0 +1,125 @@
+}
+
+#[cfg(test)]
+mod tests {
```
Please give this a more descriptive name; when this fails when running all tests in Prusti, it just says "failed running tests" or something similar.

fpoli replied:
What's the error message? I think that the convention is to always use `mod test { .. }`, because having unique names for each test module sounds difficult to enforce.

dario23 replied:
@fpoli i can remove the `s` if you want
@Aurel300  the error seems pretty clear to me; it includes the full module path (except the crate name, but that's not something i can fix with the module name here):

```
failures:
    vir::vir_macro::tests::forall

test result: FAILED. 9 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.04s
```

Aurel300 replied:
Ok, this is fine then. I remember dealing with failures that just said `tests` instead of the full module path; I guess this happened when the `tests` module was top-level in its crate, since the crate name is not shown.

dario23 committed:
Tidy up vir! macro

 - use builder functions on Expr and Stmt where available

dario23 committed:
Allow multiple trigger sets in forall

dario23 committed:
Use $crate macro meta-variable instead of crate keyword

The difference (in our use of it) is that $crate is expanded at the
macro *definition* site while the keyword `crate` refers to the current
crate at macro *expansion*/*usage* site.

dario23 committed:
export vir macro

dario23 committed:
add more binops to vir! macro

dario23 committed:
Export vir_type! and vir_local!

they are used in the definition of vir!
